src/Microsoft.AspNetCore.Server.IntegrationTesting/Common/TestUriHelper.cs

```diff
@@ -11,39 +11,68 @@ public static class TestUriHelper
     {
         public static Uri BuildTestUri()
         {
-            return new UriBuilder("http", "localhost", GetNextPort()).Uri;
+            return BuildTestUri(null);
         }
 
         public static Uri BuildTestUri(string hint)
+        {
+            return BuildTestUri(hint, statusMessagesEnabled: false);
+        }
+
+        public static Uri BuildTestUri(string hint, bool statusMessagesEnabled)
         {
             if (string.IsNullOrEmpty(hint))
             {
-                return BuildTestUri();
+                if (statusMessagesEnabled)
+                {
+                    // Most functional tests use this codepath and should directly bind to dynamic port "0" and scrape
+                    // the assigned port from the status message, which should be 100% reliable since the port is bound
+                    // once and never released.  Binding to dynamic port "0" on "localhost" (both IPv4 and IPv6) is not
+                    // supported, so the port is only bound on "127.0.0.1" (IPv4).  If a test explicitly requires IPv6,
+                    // it should provide a hint URL with "localhost" (IPv4 and IPv6) or "[::1]" (IPv6-only).
+                    return new UriBuilder("http", "127.0.0.1", 0).Uri;
+                }
+                else
+                {
+                    // If status messages are disabled, there is no status message from which to scrape the assigned port,
+                    // so the less reliable GetNextPort() must be used.  The port is bound on "localhost" (both IPv4 and IPv6),
+                    // since this is supported when using a specific (non-zero) port.
+                    return new UriBuilder("http", "localhost", GetNextPort()).Uri;
```

Comment:
Adding retries is possible but difficult.  The AddressInUse exception doesn't come from `GetNextPort()`, it comes from the application started in an external process.  So the retry logic would need to be here:

https://github.com/aspnet/Hosting/blob/dev/src/Microsoft.AspNetCore.Server.IntegrationTesting/Deployers/SelfHostDeployer.cs#L47

I think we'd need to try to start the app, and if it fails to start we'd need to scrape the output for AddressInUseException, and then retry on a different port.  We'd need to be careful to *not* retry if it failed for any other reason.  We'd also need to be careful to only retry if `BuildTestUri()` used `GetNextPort()`, since there's no point retrying if the test is using a hardcoded port.

Overall, this seems risky and error-prone, and it might cause more problems than it solves.  I think we should wait and see how many flaky tests remain after changing the default to use a dynamic port.

Reply:
Yea it's just another enhancement we could add. Doesn't have to be in this PR which is why I approved it already.